File: apps/api/app/schemas/academics.py
import uuid

from pydantic import BaseModel, ConfigDict, Field


class PaginationMeta(BaseModel):
    page: int
    page_size: int
    total: int


# ---------------------------------------------------------------------------
# Department
# ---------------------------------------------------------------------------


class DepartmentBase(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    code: str = Field(min_length=1, max_length=32)
    campus_id: uuid.UUID | None = None


class DepartmentCreate(DepartmentBase):
    pass


class DepartmentUpdate(BaseModel):
    name: str | None = None
    code: str | None = None
    campus_id: uuid.UUID | None = None


class DepartmentRead(DepartmentBase):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID


class DepartmentListResponse(BaseModel):
    data: list[DepartmentRead]
    meta: PaginationMeta


# ---------------------------------------------------------------------------
# Subject
# ---------------------------------------------------------------------------


class SubjectBase(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    code: str = Field(min_length=1, max_length=32)
    department_id: uuid.UUID
    credits: int = Field(ge=0)


class SubjectCreate(SubjectBase):
    pass


class SubjectUpdate(BaseModel):
    name: str | None = None
    code: str | None = None
    department_id: uuid.UUID | None = None
    credits: int | None = Field(default=None, ge=0)


class SubjectRead(SubjectBase):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID


class SubjectListResponse(BaseModel):
    data: list[SubjectRead]
    meta: PaginationMeta


# ---------------------------------------------------------------------------
# Program
# ---------------------------------------------------------------------------


class ProgramBase(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    code: str = Field(min_length=1, max_length=32)
    department_id: uuid.UUID
    degree_type: str = Field(pattern="^(undergraduate|postgraduate|diploma)$")


class ProgramCreate(ProgramBase):
    pass


class ProgramRead(ProgramBase):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID


class ProgramListResponse(BaseModel):
    data: list[ProgramRead]
    meta: PaginationMeta


# ---------------------------------------------------------------------------
# Academic Year
# ---------------------------------------------------------------------------


class AcademicYearBase(BaseModel):
    name: str = Field(min_length=1, max_length=32)
    start_date: str
    end_date: str
    is_current: bool = False


class AcademicYearCreate(AcademicYearBase):
    pass


class AcademicYearRead(AcademicYearBase):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID


class AcademicYearListResponse(BaseModel):
    data: list[AcademicYearRead]
    meta: PaginationMeta


# ---------------------------------------------------------------------------
# Semester
# ---------------------------------------------------------------------------


class SemesterBase(BaseModel):
    name: str = Field(min_length=1, max_length=64)
    academic_year_id: uuid.UUID
    start_date: str
    end_date: str
    is_current: bool = False


class SemesterCreate(SemesterBase):
    pass


class SemesterRead(SemesterBase):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID


class SemesterListResponse(BaseModel):
    data: list[SemesterRead]
    meta: PaginationMeta


# ---------------------------------------------------------------------------
# Section
# ---------------------------------------------------------------------------


class SectionBase(BaseModel):
    name: str = Field(min_length=1, max_length=64)
    program_id: uuid.UUID
    semester_id: uuid.UUID
    capacity: int = Field(ge=0)
    class_teacher_id: uuid.UUID | None = None


class SectionCreate(SectionBase):
    pass


class SectionUpdate(BaseModel):
    name: str | None = None
    capacity: int | None = Field(default=None, ge=0)
    class_teacher_id: uuid.UUID | None = None


class SectionRead(SectionBase):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID


class SectionListResponse(BaseModel):
    data: list[SectionRead]
    meta: PaginationMeta

File: apps/web/src/types/academics.ts
export interface Department {
  id: string;
  name: string;
  code: string;
  campus_id: string | null;
}

export interface DepartmentListResponse {
  data: Department[];
  meta: { page: number; page_size: number; total: number };
}

export interface DepartmentCreateInput {
  name: string;
  code: string;
  campus_id?: string | null;
}

export type DepartmentUpdateInput = Partial<DepartmentCreateInput>;

export interface Subject {
  id: string;
  name: string;
  code: string;
  department_id: string;
  credits: number;
}

export interface SubjectListResponse {
  data: Subject[];
  meta: { page: number; page_size: number; total: number };
}

export interface SubjectCreateInput {
  name: string;
  code: string;
  department_id: string;
  credits: number;
}

export type SubjectUpdateInput = Partial<SubjectCreateInput>;

export interface Program {
  id: string;
  name: string;
  code: string;
  department_id: string;
  degree_type: "undergraduate" | "postgraduate" | "diploma";
}

export interface ProgramListResponse {
  data: Program[];
  meta: { page: number; page_size: number; total: number };
}

export interface ProgramCreateInput {
  name: string;
  code: string;
  department_id: string;
  degree_type: Program["degree_type"];
}

export type ProgramUpdateInput = Partial<Pick<ProgramCreateInput, "name" | "code" | "degree_type">>;

export interface AcademicYear {
  id: string;
  name: string;
  start_date: string;
  end_date: string;
  is_current: boolean;
}

export interface AcademicYearListResponse {
  data: AcademicYear[];
  meta: { page: number; page_size: number; total: number };
}

export interface AcademicYearCreateInput {
  name: string;
  start_date: string;
  end_date: string;
  is_current?: boolean;
}

export type AcademicYearUpdateInput = Partial<AcademicYearCreateInput>;

export interface Semester {
  id: string;
  name: string;
  academic_year_id: string;
  start_date: string;
  end_date: string;
  is_current: boolean;
}

export interface SemesterListResponse {
  data: Semester[];
  meta: { page: number; page_size: number; total: number };
}

export interface SemesterCreateInput {
  name: string;
  academic_year_id: string;
  start_date: string;
  end_date: string;
  is_current?: boolean;
}

export type SemesterUpdateInput = Partial<
  Pick<SemesterCreateInput, "name" | "start_date" | "end_date" | "is_current">
>;

export interface Section {
  id: string;
  name: string;
  program_id: string;
  semester_id: string;
  capacity: number;
  class_teacher_id: string | null;
}

export interface SectionListResponse {
  data: Section[];
  meta: { page: number; page_size: number; total: number };
}

export interface SectionCreateInput {
  name: string;
  program_id: string;
  semester_id: string;
  capacity: number;
  class_teacher_id?: string | null;
}

export type SectionUpdateInput = Partial<
  Pick<SectionCreateInput, "name" | "capacity" | "class_teacher_id">
>;

export interface AcademicsSummary {
  departments: number;
  programs: number;
  academic_years: number;
  semesters: number;
  sections: number;
  subjects: number;
}
